### ISSUE TITLE
Message "cannot trim 'hole': no such device in pool" on `zpool trim` when log is on raw block device instead of partition

### ISSUE THREAD

**Issue by mailinglists35** [openzfs/zfs#10906]
Cannot issue `zpool trim` after replacing the `slog` and `cache` vdevs from mbr partitions on ssd to gpt raw lvm thin volumes on ssd (same ssd).

`zpool trim` does work for a ssd-only small test pool setup on top of a lvm thin volume in the same thin pool in the lvm volume group.  It worked as well just fine before (slog/cache on partitions). It works okay even when I do a monstruous test setup of using a zvol from the small test pool as the log vdev for the main pool (adding it with zpool add log /path/dev created the classic two p1 p9 partitions on it).

Adding a second log device, this time from a partition instead of a raw lvm block device, makes trim work on both slog devices (on existing lvm raw slog and on newly added lvm partition slog)[1], but does NOT work if doing similar for cache device.

Also removing the log device and re-adding also makes trim work again on the logs device and make the error message go away, however the cache device is never trimmed with either partition or whole disk (I think this deserves a separate issue? I can manually blkdiscard it every time before pool import, until persistent l2arc is released, but it's annoying to know it doesn't work with zpool trim)

I don't know how to troubleshoot this. I don't know where "hole" word comes from and it seems to me to be a bug. I am unsure if this is reproductible - if anyone attempts to, it should try creating pool then adding log/cache on raw(unpartitioned) lvm thin volumes on ssd physical drive [2].

pool setup (ubuntu 18.04, zfs 0.8.4) - (nevermind the scrub, it behaves the same as when normal):

````
# zpool status -LPt
  pool: pool
 state: DEGRADED
status: One or more devices has experienced an unrecoverable error.  An
        attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
        using 'zpool clear' or replace the device with 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-9P
  scan: scrub in progress since Wed Sep  9 21:08:29 2020
        1.94T scanned at 178M/s, 1.07T issued at 98.1M/s, 7.15T total
        168K repaired, 14.95% done, 0 days 18:03:38 to go
config:

        NAME            STATE     READ WRITE CKSUM
        pool            DEGRADED     0     0     0
          mirror-0      DEGRADED     0     0     0
            /dev/dm-11  ONLINE       0     0     0  (trim unsupported)
            /dev/dm-10  DEGRADED     0     0    42  too many errors  (repairing)  (trim unsupported)
            /dev/dm-12  ONLINE       0     0     0  (trim unsupported)
        logs
          /dev/dm-7     ONLINE       0     0     0  (untrimmed)
        cache
          /dev/dm-8     ONLINE       0     0     0  (untrimmed)

errors: No known data errors

  pool: small
 state: ONLINE
  scan: scrub repaired 0B in 0 days 00:00:33 with 0 errors on Wed Sep  9 21:06:41 2020
config:

        NAME         STATE     READ WRITE CKSUM
        small        ONLINE       0     0     0
          /dev/dm-6  ONLINE       0     0     0  (100% trimmed, completed at Wed 09 Sep 2020 09:06:59 PM EEST)

errors: No known data errors
root@pool:~# zpool trim pool
cannot trim 'hole': no such device in pool
````

[1]
trim after adding a second slog on partition

````
        logs
          /dev/dm-18    ONLINE       0     0     0  (100% trimmed, completed at Thu 10 Sep 2020 01:25:59 AM EEST)
          /dev/dm-7     ONLINE       0     0     0  (100% trimmed, completed at Thu 10 Sep 2020 01:26:04 AM EEST)
````

[2] layers
````
root@pool:~# lsblk -Dfps /dev/dm-6
NAME                                  DISC-ALN DISC-GRAN DISC-MAX DISC-ZERO FSTYPE      LABEL UUID                                   MOUNTPOINT
/dev/mapper/poolvg-small                     0       64K      16G         0 zfs_member  small 11761402153783647478
└─/dev/mapper/poolvg-thinpool-tpool          0      512B       2G         0
  ├─/dev/mapper/poolvg-thinpool_tmeta        0      512B       2G         0
  │ └─/dev/sde2                              0      512B       2G         0 LVM2_member       68ZPB5-PsbY-JSE4-qqke-EmTI-3sk8-BjXLmd
  │   └─/dev/sde                             0      512B       2G         0
  └─/dev/mapper/poolvg-thinpool_tdata        0      512B       2G         0
    └─/dev/sde2                              0      512B       2G         0 LVM2_member       68ZPB5-PsbY-JSE4-qqke-EmTI-3sk8-BjXLmd
      └─/dev/sde                             0      512B       2G         0
root@pool:~# lsblk -Dfps /dev/dm-7
NAME                                  DISC-ALN DISC-GRAN DISC-MAX DISC-ZERO FSTYPE      LABEL UUID                                   MOUNTPOINT
/dev/mapper/poolvg-zfs.logs                  0       64K      16G         0 zfs_member  pool  1694953391382931788
└─/dev/mapper/poolvg-thinpool-tpool          0      512B       2G         0
  ├─/dev/mapper/poolvg-thinpool_tmeta        0      512B       2G         0
  │ └─/dev/sde2                              0      512B       2G         0 LVM2_member       68ZPB5-PsbY-JSE4-qqke-EmTI-3sk8-BjXLmd
  │   └─/dev/sde                             0      512B       2G         0
  └─/dev/mapper/poolvg-thinpool_tdata        0      512B       2G         0
    └─/dev/sde2                              0      512B       2G         0 LVM2_member       68ZPB5-PsbY-JSE4-qqke-EmTI-3sk8-BjXLmd
      └─/dev/sde                             0      512B       2G         0
````

**Comment by ahrens** [comment]
`hole` probably refers to a vdev id that's no longer in use.  The trim code probably needs to have a special case added to ignore these non-existent vdevs.  @behlendorf

**Comment by olidal** [comment]
Same issue here with zfs version coming with Proxmox:

```
root@bougon:~# zpool trim zpool
cannot trim 'hole': no such device in pool
```

```
root@bougon:~# uname -a
Linux bougon 5.4.65-1-pve #1 SMP PVE 5.4.65-1 (Mon, 21 Sep 2020 15:40:22 +0200) x86_64 GNU/Linux
root@bougon:~# zfs --version
zfs-0.8.4-pve2
zfs-kmod-0.8.4-pve1
```

```
root@bougon:~# zpool status
  pool: rpool
 state: ONLINE
  scan: scrub repaired 0B in 0 days 00:00:13 with 0 errors on Sun Nov  8 01:24:15 2020
config:

	NAME                                                  STATE     READ WRITE CKSUM
	rpool                                                 ONLINE       0     0     0
	  ata-PNY_CS900_120GB_SSD_PNY44191911011503AD5-part3  ONLINE       0     0     0

errors: No known data errors

  pool: zpool
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0B in 0 days 02:29:51 with 0 errors on Sun Nov  8 02:53:54 2020
config:

	NAME                                           STATE     READ WRITE CKSUM
	zpool                                          ONLINE       0     0     0
	  raidz1-0                                     ONLINE       0     0     0
	    scsi-36848f690db634200272d8b903151f8e3     ONLINE       0     0     0
	    scsi-36848f690db634200272d8bda35b6c2cb     ONLINE       0     0     0
	    scsi-36848f690db634200272d8bf63765b5ba     ONLINE       0     0     0
	    scsi-36848f690db634200272d8c0c38bac815     ONLINE       0     0     0
	logs	
	  nvme-Force_MP510_19478236000128863572-part5  ONLINE       0     0     0
	cache
	  nvme-Force_MP510_19478236000128863572-part6  ONLINE       0     0     0

errors: No known data errors
```

```
root@bougon:~# zpool list -Lv
NAME          SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
rpool          31G  1.60G  29.4G        -         -     5%     5%  1.00x    ONLINE  -
  sda3         31G  1.60G  29.4G        -         -     5%  5.14%      -  ONLINE  
zpool        10.9T  1.99T  8.92T        -         -     6%    18%  1.00x    ONLINE  -
  raidz1     10.9T  1.99T  8.92T        -         -     6%  18.2%      -  ONLINE  
    sdd          -      -      -        -         -      -      -      -  ONLINE  
    sdb          -      -      -        -         -      -      -      -  ONLINE  
    sdc          -      -      -        -         -      -      -      -  ONLINE  
    sde          -      -      -        -         -      -      -      -  ONLINE  
logs             -      -      -        -         -      -      -      -  -
  nvme0n1p5  3.75G  9.34M  3.74G        -         -     0%  0.24%      -  ONLINE  
cache            -      -      -        -         -      -      -      -  -
  nvme0n1p6   156G   135G  20.1G        -         -     0%  87.1%      -  ONLINE  
```

Here is how the pool was originally created (the disabled feature is due to compatibility issues with another system when sending/receiving):
```
root@bougon:~# zpool history zpool | head -2
History for 'zpool':
2020-10-29.16:21:13 zpool create -o feature@userobj_accounting=disabled -o ashift=12 -o listsnaps=on -o feature@lz4_compress=enabled -o feature@hole_birth=enabled -o feature@embedded_data=enabled -O compression=on zpool raidz1 /dev/disk/by-id/scsi-36848f690db634200272d8b903151f8e3 /dev/disk/by-id/scsi-36848f690db634200272d8bda35b6c2cb /dev/disk/by-id/scsi-36848f690db634200272d8bf63765b5ba /dev/disk/by-id/scsi-36848f690db634200272d8c0c38bac815 cache /dev/disk/by-id/ata-PNY_CS900_120GB_SSD_PNY44191911011503AD5-part7 log /dev/disk/by-id/ata-PNY_CS900_120GB_SSD_PNY44191911011503AD5-part5
```

I tried removing both logs and cache SSD vdevs, which turned to be quite an adventure, because lots of processes ended up stuck in a blocking write operation (D state in ps). I had to reboot twice the server and only succeeded after shutting down most VMs and CTs. 

After removal, funnily, the error was still here, but it disappeared when I re-added the two vdevs as cache and logs.

Bottom-line: it is good to know that you can always reboot (power on-off) a ZFS system in case of trouble without the fear of reaching a fatal inconsistent state!

**Comment by behlendorf** [comment]
This issue was fixed by #10372 but not backported to the 0.8 branch, the fix will be in the 2.0 release.  As a workaround for now you can explicitly specify the vdevs to trim, e.g. `zpool trim rpool nvme0n1p5 nvme0n1p6`, or simply ignore the warning.

**Comment by olidal** [comment]
Thanks.

Le 17 nov. 2020 à 19:45, à 19:45, Brian Behlendorf <notifications@github.com> a écrit:

**Comment by mclad** [comment]
Problem still with us. Single vdev root pool on an SSD
Strangely boot pool on same SSD neither with cache or slog has no problem - i'm thinking something in the pool properties that are enabled on the root pool and disabled on the bootpool

<pre><font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool initialize lpool
cannot initialize &apos;hole&apos;: no such device in pool
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool trim lpool
cannot trim &apos;hole&apos;: no such device in pool
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool scrub lpool
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool status -tv lpool
  pool: lpool
 state: ONLINE
  scan: scrub in progress since Sun May  9 18:43:19 2021
	46.0G scanned at 2.00G/s, 1.91M issued at 84.9K/s, 116G total
	0B repaired, 0.00% done, no estimated completion time
config:

	NAME                                              STATE     READ WRITE CKSUM
	lpool                                             ONLINE       0     0     0
	  ata-HFS512G39TND-N210A_EJ72N136611202G5E-part9  ONLINE       0     0     0  (100% trimmed, completed at Sun 09 May 2021 06:13:46 PM AWST)

errors: No known data errors
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool --version lpool
zfs-0.8.5-pve1
zfs-kmod-2.0.4-pve1
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zls -r lpool
NAME                                                USED  USEDSNAP     REFER  CANMOUNT  MOUNTED  MOUNTPOINT
lpool                                               126G        8K       96K       off       no  /
lpool/ROOT                                         99.0G        8K       96K       off       no  none
lpool/ROOT/arcoA_base01                            9.45G      497M     8.96G        on      yes  /zab
lpool/ROOT/arcoA_zevm01                            16.2G      776K     16.1G    noauto       no  /
lpool/ROOT/arcoA_zevm01/var                         110M        0B       96K       off       no  /var
lpool/ROOT/arcoA_zevm01/var/lib                     110M       80K     10.6M    noauto       no  /var/lib
lpool/ROOT/arcoA_zevm01/var/lib/AccountsService     172K       64K      108K    noauto       no  /var/lib/AccountsService
lpool/ROOT/arcoA_zevm01/var/lib/NetworkManager      228K       88K      140K    noauto       no  /var/lib/NetworkManager
lpool/ROOT/arcoA_zevm01/var/lib/pacman             98.8M       64K     98.7M    noauto       no  /var/lib/pacman
lpool/ROOT/arcoA_zevm01/var/log                     408K       80K      328K    noauto       no  /var/log
lpool/ROOT/buster_base01                           1.48G      888K     1.40G    noauto       no  /
lpool/ROOT/buster_base01/usr                        256K        0B       96K       off       no  /usr
lpool/ROOT/buster_base01/usr/local                  160K       64K       96K    noauto       no  /usr/local
lpool/ROOT/buster_base01/var                       79.9M        0B       96K       off       no  /var
lpool/ROOT/buster_base01/var/lib                   63.0M       72K     8.73M    noauto       no  /var/lib
lpool/ROOT/buster_base01/var/lib/AccountsService     96K        0B       96K    noauto       no  /var/lib/AccountsService
lpool/ROOT/buster_base01/var/lib/NetworkManager     120K        0B      120K    noauto       no  /var/lib/NetworkManager
lpool/ROOT/buster_base01/var/lib/apt               37.0M       64K     36.9M    noauto       no  /var/lib/apt
lpool/ROOT/buster_base01/var/lib/dpkg              17.1M       64K     17.0M    noauto       no  /var/lib/dpkg
lpool/ROOT/buster_base01/var/log                   16.8M       64K     16.7M    noauto       no  /var/log
lpool/ROOT/focal_kvmq01                            2.37G     1.66G     3.87G    noauto       no  /
lpool/ROOT/focal_kvmq01/var                         104K        0B       96K       off       no  /var
lpool/ROOT/focal_kvmq01/var/log                      96K        0B       96K    noauto       no  /var/log
lpool/ROOT/focal_uni2020                           9.72G      999M     8.75G    noauto       no  /
lpool/ROOT/groovy_base01                           4.55G      420M     4.66G    noauto       no  /pv/alt/newroot/
lpool/ROOT/groovy_base01/var                       47.1M        0B       96K       off       no  /pv/alt/newroot//var
lpool/ROOT/groovy_base01/var/log                   47.1M     6.55M     40.5M    noauto       no  /pv/alt/newroot//var/log
lpool/ROOT/groovy_host01                           19.9G     14.0G     5.85G    noauto       no  /
lpool/ROOT/groovy_host01/var                       73.5M        0B       96K       off       no  /var
lpool/ROOT/groovy_host01/var/log                   73.4M     18.6M     54.9M    noauto       no  /var/log
lpool/ROOT/hirsute_base01                          14.0G     7.03G     3.59G    noauto       no  /zhh/
lpool/ROOT/hirsute_base01/usr                      82.0M        0B       96K       off       no  /zhh//usr
lpool/ROOT/hirsute_base01/usr/local                81.9M     81.8M      128K    noauto       no  /zhh//usr/local
lpool/ROOT/hirsute_base01/var                      3.27G       80K     1.21M       off       no  /zhh//var
lpool/ROOT/hirsute_base01/var/lib                  2.59G      318M     1.26G    noauto       no  /zhh//var/lib
lpool/ROOT/hirsute_base01/var/lib/AccountsService   776K      664K      112K    noauto       no  /zhh//var/lib/AccountsService
lpool/ROOT/hirsute_base01/var/lib/NetworkManager   1.53M     1.37M      164K    noauto       no  /zhh//var/lib/NetworkManager
lpool/ROOT/hirsute_base01/var/lib/apt               891M      809M     82.2M    noauto       no  /zhh//var/lib/apt
lpool/ROOT/hirsute_base01/var/lib/dpkg              158M      124M     33.4M    noauto       no  /zhh//var/lib/dpkg
lpool/ROOT/hirsute_base01/var/log                   692M      108M      584M    noauto       no  /zhh//var/log
lpool/ROOT/hirsute_base01/var/snap                  772K      644K      128K    noauto       no  /zhh//var/snap
lpool/ROOT/pve-1                                   1.79G      107M     1.69G    noauto       no  /pv/
lpool/ROOT/pve10_gnom01                            15.8G     9.22G     6.54G        on      yes  /
lpool/USERDATA                                     8.70G        8K       96K       off       no  none
lpool/USERDATA/arch                                1.10G      636M      495M        on      yes  /zab/home
lpool/USERDATA/common-files                        2.88G      184M     2.70G        on      yes  /media/files
lpool/USERDATA/debian                               148M     73.7M     73.9M    noauto       no  /home
lpool/USERDATA/pve-tower                           1.50G      470M     1.04G        on      yes  /home
lpool/USERDATA/ubuntu                              3.07G      723M     2.36G    noauto       no  /home
lpool/USERDATA/zsys_skel                            224K       64K      160K    noauto       no  /home
lpool/data                                         17.0G      136K     6.66G       off       no  /lpool/data
lpool/data/vm-100-disk-0                           10.3G        8K       56K         -        -  -
lpool/data/vm-100-disk-1                           3.06M        8K       56K         -        -  -
lpool/opt                                          1.47G      723M      786M        on      yes  /opt
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool trim bpool
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool status -tv bpool
  pool: bpool
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using &apos;zpool upgrade&apos;. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0B in 0 days 00:00:09 with 0 errors on Tue Apr 20 15:06:52 2021
config:

	NAME                                              STATE     READ WRITE CKSUM
	bpool                                             ONLINE       0     0     0
	  ata-HFS512G39TND-N210A_EJ72N136611202G5E-part8  ONLINE       0     0     0  (100% trimmed, completed at Sun 09 May 2021 06:48:34 PM AWST)

errors: No known data errors
</pre>

**Comment by mclad** [comment]
So here are the pool properties which is the only difference. Let me know if further information would assist.

<pre><font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool get all lpool | egrep &apos; feature\@&apos;
lpool <font color="#C01C28"><b> feature@</b></font>async_destroy          enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>empty_bpobj            active                         local
lpool <font color="#C01C28"><b> feature@</b></font>lz4_compress           active                         local
lpool <font color="#C01C28"><b> feature@</b></font>multi_vdev_crash_dump  enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>spacemap_histogram     active                         local
lpool <font color="#C01C28"><b> feature@</b></font>enabled_txg            active                         local
lpool <font color="#C01C28"><b> feature@</b></font>hole_birth             active                         local
lpool <font color="#C01C28"><b> feature@</b></font>extensible_dataset     active                         local
lpool <font color="#C01C28"><b> feature@</b></font>embedded_data          active                         local
lpool <font color="#C01C28"><b> feature@</b></font>bookmarks              enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>filesystem_limits      enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>large_blocks           enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>large_dnode            active                         local
lpool <font color="#C01C28"><b> feature@</b></font>sha512                 enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>skein                  enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>edonr                  enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>userobj_accounting     active                         local
lpool <font color="#C01C28"><b> feature@</b></font>encryption             enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>project_quota          active                         local
lpool <font color="#C01C28"><b> feature@</b></font>device_removal         enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>obsolete_counts        enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>zpool_checkpoint       enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>spacemap_v2            active                         local
lpool <font color="#C01C28"><b> feature@</b></font>allocation_classes     enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>resilver_defer         enabled                        local
lpool <font color="#C01C28"><b> feature@</b></font>bookmark_v2            enabled                        local
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool get all bpool | egrep &apos; feature\@&apos;
bpool <font color="#C01C28"><b> feature@</b></font>async_destroy          enabled                        local
bpool <font color="#C01C28"><b> feature@</b></font>empty_bpobj            active                         local
bpool <font color="#C01C28"><b> feature@</b></font>lz4_compress           active                         local
bpool <font color="#C01C28"><b> feature@</b></font>multi_vdev_crash_dump  disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>spacemap_histogram     active                         local
bpool <font color="#C01C28"><b> feature@</b></font>enabled_txg            active                         local
bpool <font color="#C01C28"><b> feature@</b></font>hole_birth             active                         local
bpool <font color="#C01C28"><b> feature@</b></font>extensible_dataset     enabled                        local
bpool <font color="#C01C28"><b> feature@</b></font>embedded_data          active                         local
bpool <font color="#C01C28"><b> feature@</b></font>bookmarks              enabled                        local
bpool <font color="#C01C28"><b> feature@</b></font>filesystem_limits      enabled                        local
bpool <font color="#C01C28"><b> feature@</b></font>large_blocks           enabled                        local
bpool <font color="#C01C28"><b> feature@</b></font>large_dnode            disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>sha512                 disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>skein                  disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>edonr                  disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>userobj_accounting     disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>encryption             disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>project_quota          disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>device_removal         disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>obsolete_counts        disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>zpool_checkpoint       disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>spacemap_v2            disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>allocation_classes     disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>resilver_defer         disabled                       local
bpool <font color="#C01C28"><b> feature@</b></font>bookmark_v2            disabled                       local
</pre>

**Comment by mclad** [comment]
Could this have something to do with it? So if I do a whatsit, zpool labelclear device, and then trim works... no dice i'm afraid. After a restart with the destroyed pool labelcleared will report if any difference.
<pre><font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool import -d /dev/disk/by-id
no pools available to import
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool import -d /dev/disk/by-id -DfN
   pool: zfstore
     id: 10998968980121345450
  state: FAULTED (DESTROYED)
 status: The pool metadata is corrupted.
 action: The pool cannot be imported due to damaged devices or data.
	The pool was destroyed, but can be imported using the &apos;-Df&apos; flags.
   see: http://zfsonlinux.org/msg/ZFS-8000-72
 config:

	zfstore                                            FAULTED  corrupted data
	logs	
	  ata-HFS512G39TND-N210A_EJ72N136611202G5E-part14  ONLINE
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool labelclear -f ata-HFS512G39TND-N210A_EJ72N136611202G5E-part14
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool import -d /dev/disk/by-id -DfN
no pools available to import
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~</b></font># zpool trim lpool
cannot trim &apos;hole&apos;: no such device in pool
</pre>

**Comment by mclad** [comment]
SOLVED by upgrading system - but no upgrade commands given to the pool so it is unchanged, just zfs version This still shouldn't have been happening on 0.85 buy anyway, by the by.
<pre><font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~/Templates</b></font># zpool --version lpool
zfs-2.0.4-pve1
zfs-kmod-2.0.4-pve1
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~/Templates</b></font># zpool trim lpool
<font color="#26A269"><b>root@pve-tower</b></font>:<font color="#12488B"><b>~/Templates</b></font># zpool status -vt lpool
  pool: lpool
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using &apos;zpool upgrade&apos;. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0B in 00:08:19 with 0 errors on Sun May  9 18:51:38 2021
config:

	NAME                                              STATE     READ WRITE CKSUM
	lpool                                             ONLINE       0     0     0
	  ata-HFS512G39TND-N210A_EJ72N136611202G5E-part9  ONLINE       0     0     0  (45% trimmed, started at Mon 10 May 2021 03:51:15 AM AWST)

errors: No known data errors
</pre>
